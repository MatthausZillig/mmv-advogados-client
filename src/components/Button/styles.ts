import styled, { css, DefaultTheme } from 'styled-components'
import { ButtonProps } from '.'

type WrapperProps = { hasIcon: boolean } & Pick<
  ButtonProps,
  'size' | 'fullWidth' | 'withBorder'
>

const WrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xxsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  widthIcon: (theme: DefaultTheme) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${theme.colors.white};
    background: none;
    box-shadow: none;
    :hover svg {
      margin-left: 1.2rem;
      transition: margin-left 100ms;
    }
    :hover {
      background: none;
    }
    :active {
      background-color: none;
      background-size: 100%;
    }
    svg {
      width: 1.5rem;
      margin-left: ${theme.spacings.xsmall};
    }
  `,
  withBorder: (theme: DefaultTheme) => css`
    border: 1px solid ${theme.colors.primary};
    color: ${theme.colors.primary};
    background: none;
    box-shadow: none;
    :hover {
      background: rgba(25, 118, 210, 0.04);
      border: 1px solid ${theme.colors.primary};
      transition: background 100ms ease;
    }
    :active {
      background-color: ${theme.colors.primary};
      background-size: 100%;
      transition: border 0s;
      transition: background 100s ease;
      border: 1px solid ${theme.colors.primary};
      color: ${theme.colors.white};
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, fullWidth, hasIcon, withBorder }) => css`
    background: ${theme.colors.primary};
    color: ${theme.colors.white};
    border: 0;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    background-position: center;
    transition: background 0.8s;
    box-shadow: 0px 2px 5px -2px rgba(0, 0, 0, 0.52);
    cursor: pointer;
    outline: none;
    text-decoration: none;
    :hover {
      background: ${theme.colors.secondary};
        radial-gradient(circle, transparent 1%, ${theme.colors.darkBlue} 1%)
        center/15000%;
    }
    :active {
      background-color: ${theme.colors.darkBlue};
      background-size: 100%;
      transition: background 0s;
    }
    ${!!size && WrapperModifiers[size](theme)};
    ${!!fullWidth && WrapperModifiers.fullWidth()};
    ${hasIcon && WrapperModifiers.widthIcon(theme)};
    ${withBorder && WrapperModifiers.withBorder(theme)};
  `}
`
