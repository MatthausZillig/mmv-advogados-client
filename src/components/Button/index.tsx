import * as S from './styles'

export type ButtonProps = {
  children?: React.ReactNode
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  withBorder?: boolean
  icon?: JSX.Element
  onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void
}

const Button = ({
  children,
  icon,
  size = 'medium',
  fullWidth = false,
  ...props
}: ButtonProps) => (
  <S.Wrapper fullWidth={fullWidth} size={size} hasIcon={!!icon} {...props}>
    {!!children && <span>{children}</span>}
    {!!icon && icon}
  </S.Wrapper>
)

export default Button
