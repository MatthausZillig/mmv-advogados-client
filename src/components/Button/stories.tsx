import { Story, Meta } from '@storybook/react/types-6-0'
import { ArrowRight } from '@styled-icons/bootstrap/ArrowRight'

import Button from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Default: Story = (args) => <Button {...args} />
export const withIcon: Story = (args) => <Button {...args} />
export const withBorder: Story = (args) => <Button {...args} />

Default.args = {
  children: 'Veja mais'
}

withIcon.args = {
  size: 'small',
  children: 'Veja mais',
  icon: <ArrowRight />
}

withBorder.args = {
  size: 'medium',
  children: 'Veja mais',
  withBorder: true
}
