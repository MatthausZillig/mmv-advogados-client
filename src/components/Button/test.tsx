import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button from '.'

describe('<Button />', () => {
  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Buy now</Button>)
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.6rem'
    })
  })
  it('should render a fullWidth version', () => {
    renderWithTheme(<Button fullWidth>Buy now</Button>)
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      width: '100%'
    })
  })
  it('should render an icon version', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Buy now</Button>
    )
    expect(screen.getByText(/buy now/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })
  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Buy now
      </Button>
    )

    expect(screen.getByRole('link', { name: /buy now/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
