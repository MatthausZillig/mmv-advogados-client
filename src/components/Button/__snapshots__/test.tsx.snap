// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a medium button by default 1`] = `
.c0 {
  background: linear-gradient(180deg,#3d7afd 0%,#7ca2ef);
  color: #FAFAFA;
  border: 0;
  border-radius: 0.4rem;
  padding: 0.8rem;
  cursor: pointer;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

<button
  class="c0"
>
  <span>
    Buy now
  </span>
</button>
`;
